### PR TITLE
Added GLFW specific functionality from GLWindow.jl

### PR DESCRIPTION
**Pull request by louisponet**
- moved functions from `glfw3.jl` to `functions.jl`
- moved types from `glfw3.jl` to `types.jl`
- moved globals from `glfw3.jl` to `constants.jl`
- move `gl_createcontext` out of `GLWindow.jl/screen.jl` to a constructor for `Window`. 
- Added a lot of `Window` related functionality from `GLWindow.jl/screen.jl,types.jl,core.jl` to `types.jl`. 
- Added `MonitorProperties` from `GLWindow.jl/types.jl` to `types.jl`
- Changed `MonitorProperties` so it doesn't depend on `GeometryTypes: Vec` hit in performance shouldn't matter imo.
- Added general `GLFW.jl` related functionality from `GLWindow/screen.jl,core.jl` to `extensions.jl`

Tests work, also ran the tests for the [pullrequest](https://github.com/JuliaGL/GLWindow.jl/pull/55) in `GLWindow.jl`

General Idea: Firstly, splitting pure `GLFW.jl` related functionality from `GLWindow.jl` will allow it to be less backend specific, and act as a interface to different `OpenGL` window/context providers. This is in light of hopefully seperating Windowing functionality in 'GLWindow' from `GLAbstraction.jl` and provide more of an `API` experience. 
Secondly, I understand this makes the package no-longer a barebones wrap around the `GLFW` library, but was it really purely that to begin with, and would people not use the added functionality that only depends on `GLFW` if it is there?

### CODE REVIEW

**Comment by louisponet** [comment]
Not sure if this improved things. I'm sorry, should I create a complete new PR, without the filechanges?

**Comment by jayschwa** [comment]
Thanks for dialing back the file reorg. It will make the other changes easier to digest.

File: src/glfw3.jl [hunk 1]
```diff
@@ -473,3 +607,105 @@ GetProcAddress(procname::AbstractString) = ccall((:glfwGetProcAddress, lib), Ptr
...
+
+#came from GLWindow/core.jl
+"""
+Returns the monitor resolution of the primary monitor.
+"""
+function primarymonitorresolution()
+    props = MonitorProperties(GetPrimaryMonitor())
+    w,h = props.videomode.width, props.videomode.height
+    # Vec(Int(w),Int(h))
+    Vector([Int(w),Int(h)])
```
Comment (SimonDanisch) [inline src/glfw3.jl]: Just return `(Int(w), Int(h))`

File: src/glfw3.jl [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#************************************************************************
+************************************************************************
```
Comment (jayschwa) [inline src/glfw3.jl]: Fix comment by re-adding `#`.

File: src/glfw3.jl [hunk 3]
```diff
@@ -473,3 +607,105 @@ GetProcAddress(procname::AbstractString) = ccall((:glfwGetProcAddress, lib), Ptr
...
+full_screen_usage_message() = """
+Keyword arg fullscreen accepts:
+    Integer: The number of the Monitor to Select
+    Bool: if true, primary monitor gets fullscreen, false no fullscren (default)
+    GLFW.Monitor: Fullscreens on the passed monitor
+"""
+
+function poll_glfw()
+    PollEvents()
+end
```
Comment (jayschwa) [inline src/glfw3.jl]: Remove this function. It is not adding value.

File: src/glfw3.jl [hunk 4]
```diff
@@ -473,3 +607,105 @@ GetProcAddress(procname::AbstractString) = ccall((:glfwGetProcAddress, lib), Ptr
...
+    ]
+end
+
+#question: what is this exactly?
+full_screen_usage_message() = """
+Keyword arg fullscreen accepts:
+    Integer: The number of the Monitor to Select
+    Bool: if true, primary monitor gets fullscreen, false no fullscren (default)
+    GLFW.Monitor: Fullscreens on the passed monitor
+"""
```
Comment (jayschwa) [inline src/glfw3.jl]: Remove this function. It appears to be unused.

File: src/glfw3.jl [hunk 5]
```diff
@@ -473,3 +607,105 @@ GetProcAddress(procname::AbstractString) = ccall((:glfwGetProcAddress, lib), Ptr
...
+#Came from: GLWindow/events.jl
+function to_arrow_symbol(button_set)
+    for b in button_set
+        KEY_RIGHT == b && return :right
+        KEY_LEFT  == b && return :left
+        KEY_DOWN  == b && return :down
+        KEY_UP    == b && return :up
+    end
+    return :nothing
+end
```
Comment (jayschwa) [inline src/glfw3.jl]: Remove this function. It appears to be unused.

File: src/glfw3.jl [hunk 6]
```diff
@@ -269,15 +269,116 @@ end
...
+    if fullscreen
+        SetKeyCallback(window, (_1, button, _2, _3, _4) -> begin
+            button == KEY_ESCAPE && make_windowed!(window)
+        end)
+        make_fullscreen!(window, monitor)
+    end
+
+    MakeContextCurrent(window)
+
+    debugging && glDebugMessageCallbackARB(_openglerrorcallback, C_NULL)
```
Comment (jayschwa) [inline src/glfw3.jl]: `glDebugMessageCallbackARB` appears to be undefined. This line might need to be removed.

File: src/glfw3.jl [hunk 7]
```diff
@@ -269,15 +269,116 @@ end
...
+    vidmodes = GetVideoModes(monitor)[end]
+    SetWindowMonitor(window, monitor, 0, 0, vidmodes.width, vidmodes.height, GLFW.DONT_CARE)
+    return
+end
+
+function swapbuffers(window::Window)
+    window.handle == C_NULL && return
+    SwapBuffers(window)
+    return
+end
```
Comment (jayschwa) [inline src/glfw3.jl]: Remove this function. It's not adding enough value in my opinion.

File: src/glfw3.jl [hunk 8]
```diff
@@ -473,3 +607,105 @@ GetProcAddress(procname::AbstractString) = ccall((:glfwGetProcAddress, lib), Ptr
...
+"""
+Takes a Window and registers a list of callback functions.
+Returns a Dict{Symbol, Any}(name_of_callback => signal)
+"""
+function register_callbacks(window::Window, callbacks::Vector{Function})
+    tmp = map(callbacks) do f
+        (Symbol(last(split(string(f),"."))), f(window))
+    end
+    Dict{Symbol, Any}(tmp)
+end
```
Comment (jayschwa) [inline src/glfw3.jl]: Remove this function (for now). I think it's a little too magical.

File: src/glfw3.jl [hunk 9]
```diff
@@ -269,15 +269,116 @@ end
...
 Base.cconvert(::Type{Window}, handle::WindowHandle) = ccall( (:glfwGetWindowUserPointer, lib), Ref{Window}, (WindowHandle,), handle)
 Base.hash(window::Window, h::UInt64) = hash(window.handle, h)
+function Base.isopen(window::Window)
+    window.handle == C_NULL && return false
+    !WindowShouldClose(window)
+end
+
+function Base.resize!(x::Window, w::Integer, h::Integer)
+    SetWindowSize(x, w, h)
+end
```
Comment (jayschwa) [inline src/glfw3.jl]: Remove this function (for now). The documented description of `resize!` is for collections, so using it here seems a bit off. The alternative (calling `SetWindowSize`) isn't any longer.

File: src/glfw3.jl [hunk 10]
```diff
@@ -269,15 +269,116 @@ end
...
+
+
 import Base.==; ==(x::Window, y::Window) = (x.handle == y.handle)
 Base.cconvert(::Type{WindowHandle}, window::Window) = window.handle
 Base.cconvert(::Type{Window}, handle::WindowHandle) = ccall( (:glfwGetWindowUserPointer, lib), Ref{Window}, (WindowHandle,), handle)
 Base.hash(window::Window, h::UInt64) = hash(window.handle, h)
+function Base.isopen(window::Window)
+    window.handle == C_NULL && return false
+    !WindowShouldClose(window)
+end
```
Comment (jayschwa) [inline src/glfw3.jl]: Remove this function (for now). "is open" and "should not close" are two different concepts in my opinion, and the alternative can still be done in one line.

**Comment by louisponet** [comment]
I'm not against any tweaking on your part I'm sure you know better what to do with this package than I do. If you want I can also delete all the annotations I made, like where blocks of code originated from.

File: .gitignore [hunk 11]
```diff
@@ -1,2 +1,3 @@
 deps/*
 !deps/build.jl
+
```
Comment (jayschwa) [inline .gitignore]: Remove this new line.

File: src/glfw3.jl [hunk 12]
```diff
@@ -269,16 +269,103 @@ end
 Base.show(io::IO, m::Monitor) = write(io, "Monitor($(m.handle == C_NULL ? m.handle : GetMonitorName(m)))")
 
 const WindowHandle = Ptr{Void}
+
```
Comment (jayschwa) [inline src/glfw3.jl]: Remove this new line.

File: src/glfw3.jl [hunk 13]
```diff
@@ -298,6 +385,39 @@ immutable GLFWError <: Exception
 end
 Base.showerror(io::IO, e::GLFWError) = print(io, "GLFWError ($(e.code)): ", e.description)
 
+#Came from GLWindow.jl/types.jl
+struct MonitorProperties
```
Comment (jayschwa) [inline src/glfw3.jl]: Change `struct` to `immutable` for Julia 0.5 compatibility.

**Comment by jayschwa** [comment]
Also, not sure if you care, but GitHub is setting your name to "Unknown" as the author. I believe that can be changed under GitHub settings.

File: src/glfw3.jl (hunk 13 above)
Comment (jayschwa) [reply src/glfw3.jl]: This request was made irrelevant by #121. However, I couldn't get GitHub to resolve merges through it's UI, so I did it manually and push it to #122.

**Comment by jayschwa** [comment]
Merged via #122. Thanks for the contribution!

**Comment by louisponet** [comment]
Very Good, thanks to you too!

**Pull request closed without merging**